sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/StorageBlockingSink.java

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.storage.blob.implementation.util;
+
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Sinks;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * This is a lightweight wrapper around Sinks.Many to implement a Sink backed by a LinkedBlockingQueue, effectively
+ * implementing a BlockingSink.
+ */
+public class StorageBlockingSink {
+    private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+    /*
+    Note: Though this is used only by BlockBlobOutputStream, the decision was made to abstract this type out to make
+    testing easier.
+    */
+    private final Sinks.Many<ByteBuffer> writeSink;
+    private final LinkedBlockingQueue<ByteBuffer> writeLimitQueue;
+
+    /* Overrided implementation of LinkedBlockingQueue to effectively implement a true BlockingSink. */
+    private static class ProducerBlockingQueue<T> extends LinkedBlockingQueue<T> {
+        private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+        ProducerBlockingQueue(int queueSize) {
+            super(queueSize);
+        }
+
+        /* The Reactor implementation of Sinks(UnicastProcessor).tryEmitNext (see usage below) makes a call to
+        queue.offer(). The desired functionality of tryEmitNext is for it to block on backpressure until there is
+        space in the queue. Since this code is called from user thread on BlobOutputStream.write and blocking on that
+        API until data can be written is ok, this is the simplest way to get desired behavior. */
+        @Override
+        public boolean offer(T o) {
+            try {
+                super.put(o);
+                return true;
+            } catch (InterruptedException e) {
+                throw logger.logExceptionAsError(new RuntimeException(e));
```

Comment:
Here, I feel like we should propogate this back somehow, maybe we can share an error variable with StorageBlockingSink here?

Reply:
options to consider:

- when we wrap it in RuntimeException then it just bubble up through tryEmitNext call which we call from blob outputstream, where we can then handle this and propagate as IOException...
- or maybe we should just return false from offer and keep that spin - loop in outputstream to handle this case. (Assuming it's rare).

Reply:
I can't think of scenario where InterruptedException would be thrown other than somebody is trying to kill tread it's executing this instruction. So propagating runtime is probably best.